perfkitbenchmarker/providers/gcp/gce_network.py

```diff
@@ -41,6 +43,405 @@
+
+    self.ip_address = None
+    self.created = False
+    self.require_target_to_init = False
+    self.routing = None
+    self.psk = None
+
+  def ConfigureTunnel(self, tunnel_config):
+    network.BaseVPNGW.ConfigureTunnel(self, tunnel_config)
+    logging.info('Configuring Tunnel with params:')
+    logging.info(tunnel_config)
+
+    # update tunnel_config if needed
+    if self.name not in tunnel_config.endpoints:
+      logging.info('tunnel_config: This endpoint isnt registered yet... %s' % self.name)
+      tunnel_config.endpoints[self.name] = {}
+      tunnel_config.endpoints[self.name]['is_configured'] = False
+      tunnel_config.endpoints[self.name]['cidr'] = self.cidr
+      tunnel_config.endpoints[self.name]['project'] = self.project
+      tunnel_config.endpoints[self.name]['network_name'] = self.network_name
+      tunnel_config.endpoints[self.name]['region'] = self.region
+      tunnel_config.endpoints[self.name]['require_target_to_init'] = self.require_target_to_init
+
+    # attach public IP to this GW if doesnt exist
+    # and update tunnel_config if needed
+    # requires project, region, name
+    if not self.ip_address:
+      if not self.ip_resource:
+        self.ip_resource = GceIPAddress(self.project, self.region, self.name)
+        self.ip_resource.Create()
+      self.ip_address = self.ip_resource.ip_address
+    if 'ip_address' not in tunnel_config.endpoints[self.name]:
+      logging.info('tunnel_config: Configuring IP for %s' % self.name)
+      tunnel_config.endpoints[self.name]['ip_address'] = self.ip_address
+
+    # configure forwarding
+    # requires: -
+    if len(self.forwarding_rules) == 3:
+      logging.info('tunnel_config: Forwarding already configured, skipping')
+    else:
+      logging.info('tunnel_config: Setting up forwarding')
+      self._SetupForwarding(tunnel_config)
+
+    # Abort if we don't have a target info configured yet
+    if len(tunnel_config.endpoints) < 2:
+      logging.info('tunnel_config: Only found %d endpoints... waiting for target to configure' % len(tunnel_config.endpoints))
+      return
+
+    # Get target endpoint config key
+    target_endpoint = [k for k in tunnel_config.endpoints.keys() if k not in self.name][0]
+
+    # configure tunnel resources
+    # requires: target_ip_address, IKE version (default 1),
+    if 'ip_address' not in tunnel_config.endpoints[target_endpoint]:
+      logging.info('tunnel_config: Target IP needed... waiting for target to configure')
+      return
+    if not hasattr(tunnel_config, 'psk'):
+      logging.info('tunnel_config: PSK not provided... setting to runid')
+      tunnel_config.psk = 'key' + FLAGS.run_uri
+    self._SetupTunnel(tunnel_config)
+
+    # configure routing
+    # requires: next_hop_tunnel_id, target_cidr,
+    dest_cidr = tunnel_config.endpoints[target_endpoint].get('cidr')
+    if not dest_cidr or not dest_cidr.strip():
+      logging.info('tunnel_config: destination CIDR needed... waiting for target to configure')
```

Comment:
When you say "waiting for target to configure" does it mean that this method will be retried later on?  I don't see where that's done.  
You could throw an errors.Error and put a @vm_util.Retry decorator around this method if that is your intention

Reply:
Each endpoint (network.BaseVpnGW) needs to know some facts about the other endpoint (target) like ip address, ike version, ASN, etc. to build a tunnel. These facts are learned incrementally during tunnel setup and added to a shared dictionary (vpn_service.TunnelConfig) as they become available. A vpn_service.VPN instance holds references to 2 endpoint instances (the gateway_pair) and 1 tunnel_config instance. 

During setup, the VPN instance just walks the tunnel_config back and forth between endpoints ([src](https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/blob/2c277d8407add37d180113feb283bd10d7e32759/perfkitbenchmarker/vpn_service.py#L121)) until both endpoints have enough information to setup the tunnel or neither endpoint can contribute new facts and the setup fails.  The "waiting for target to configure" message just indicates that we can't continue configuring this end of the tunnel until we know more about the other gateway - not that we're blocking (or even really waiting, since the gateway just punts here and the VPN instance handles state). I see why that message is misleading. The configureTunnel() method we're in here needs to be agnostic to the other endpoint and idempotent across multiple config attempts, so 'best-effort' is preferable to retrying I think. Happy to revisit this if I didn't address your concern.